File: codingTest/Programmers/여행경로/src.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>
using namespace std;
int n;
bool dfs(vector<vector<string>> const tickets, vector<string> &route, vector<bool> &visited, string dst){
    route.push_back(dst);
    if(route.size() == n+1) return true;
    for(int i = 0; i<n; i++){
        if(!visited[i] && tickets[i][0] == dst){
            visited[i] = true;
            if(dfs(tickets, route, visited, tickets[i][1])){
                return true;
            }
            visited[i] = false;
        }
    }
    route.pop_back();
    return false;
}
vector<string> solution(vector<vector<string>> tickets) {
    vector<string> answer;
    n = tickets.size();
    sort(tickets.begin(), tickets.end());
    for(int i = 0; i < n; i++){
        vector<bool> visited(n, false);
        vector<string> route;
        if(tickets[i][0] == "ICN"){
            visited[i] = true;
            route.push_back("ICN");
            if(dfs(tickets, route, visited, tickets[i][1])){
                answer = route;
                break;
            }
            
        }
    }
    return answer;
}

File: codingTest/Programmers/여행경로/src.py
def solution(tickets):
    answer = ["ICN"]
    candidates = []
    n = len(tickets)
    visited = [False] * n

    def dfs(depart):
        if len(answer) == n + 1:
            candidates.append(list(answer))
            return
        for idx, ticket in enumerate(tickets):
            if depart == ticket[0] and not visited[idx]:
                answer.append(ticket[1])
                visited[idx] = True
                dfs(ticket[1])
                visited[idx] = False
                answer.pop()
    dfs("ICN")
    
    candidates.sort()
    answer = candidates[0]
    return answer
